common/src/main/java/io/netty/util/DomainNameMapping.java

```diff
@@ -141,4 +149,96 @@ public V map(String input) {
     public String toString() {
         return StringUtil.simpleClassName(this) + "(default: " + defaultValue + ", map: " + map + ')';
     }
+
+    /**
+     * Creates a new builder for {@link DomainNameMapping}.
+     *
+     * @param defaultValue the default value for {@link #map(String)} to return when nothing matches the input
+     */
+    public static <V> DomainMappingBuilder<V> withDefaultValue(V defaultValue) {
+        return withInitialCapacityAndDefaultValue(4, defaultValue);
+    }
+
+    /**
+     * Creates a new builder for {@link DomainNameMapping}.
+     *
+     * @param initialCapacity initial capacity for the internal map
+     * @param defaultValue    the default value for {@link #map(String)} to return when nothing matches the input
+     */
+    public static <V> DomainMappingBuilder<V> withInitialCapacityAndDefaultValue(int initialCapacity, V defaultValue) {
+        return new DomainMappingBuilder<V>(initialCapacity, defaultValue);
+    }
+
+    /**
+     * Immutable mapping from domain name pattern to its associated value object.
+     * Mapping is represented by two arrays: keys and values. Key domainNamePatterns[i] is associated with values[i].
+     *
+     * @param <V> concrete type of value objects
+     */
+    static final class ImmutableDomainNameMapping<V> extends DomainNameMapping<V> {
+        private final String[] domainNamePatterns;
+        private final V[] values;
+
+        @SuppressWarnings("unchecked")
+        ImmutableDomainNameMapping(V defaultValue, Map<String, V> map) {
+            super(/* map */null, defaultValue);
+
+            Set<Map.Entry<String, V>> mappings = map.entrySet();
+            domainNamePatterns = new String[mappings.size()];
```

Comment:
store mappings.size() in a local variable. No need to call it multiple times as this may be expensive